### PR TITLE
#1293 Implement Matrix Trace

### PR DESCRIPTION
**Pull request by cameronphchen**
As discussed in #1293, this is an attempt to implement matrix trace operator for rank 2 tensors. It only takes tensors with rank 2 (similar to the trace design in Theano). Trace for higher rank tensors requires more discussion on how we want to implement it. One approach is to calculate trace for all rank 2 sub-tensors as numpy. Another is to implement [tensor contraction](https://en.wikipedia.org/wiki/Tensor_contraction) (generalization of trace).

### CODE REVIEW

File: tensorflow/python/ops/math_ops.py [hunk 1]
```diff
@@ -920,6 +921,39 @@ def reduce_any(input_tensor, reduction_indices=None, keep_dims=False,
...
+  tf.trace(x) ==> 2
+  
+  # 'x' is [[1,2,3],
+  #         [4,5,6],
+  #         [7,8,9]]
+  tf.trace(x) ==> 15
+  ```
+
+  Args:
+    input_tensor: Rank 2 tensor.
```
Comment (martinwicke) [inline tensorflow/python/ops/math_ops.py]: Can you instead write "input_tensor: 2-D tensor." to be consistent with other ops with dimensionality constraints?

File: tensorflow/python/ops/math_ops.py [hunk 2]
```diff
@@ -920,6 +921,39 @@ def reduce_any(input_tensor, reduction_indices=None, keep_dims=False,
                                                         reduction_indices),
                            keep_dims, name=name)
 
+def trace(x, name=None):
+  """ Compte the trace of a tensor `x`.
```
Comment (martinwicke) [inline tensorflow/python/ops/math_ops.py]: Compute

File: tensorflow/python/kernel_tests/trace_op_test.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy
+import tensorflow as tf
+
+
+class GenerateIdentityTensorTest(tf.test.TestCase):
```
Comment (martinwicke) [inline tensorflow/python/kernel_tests/trace_op_test.py]: Can you fix class name?

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
Just some docs and naming nits, otherwise looks good!

**Comment by cameronphchen** [comment]
@martinwicke Thank you so much for the review! I have updated them accordingly.

File: tensorflow/python/ops/math_ops.py [hunk 4]
```diff
@@ -920,6 +921,39 @@ def reduce_any(input_tensor, reduction_indices=None, keep_dims=False,
...
+    The trace of input tensor.
+  """
+  #if not isinstance(x, (ops.Tensor, ops.IndexedSlices)):
+  #  raise TypeError("Not a Tensor or IndexedSlices: %s" % type(x))
+  with ops.op_scope([x], name, "Trace") as name: 
+    x = ops.convert_to_tensor(x, name="x")
+    if len(x.get_shape()) != 2:
+      raise ValueError("Expected a tensor with rank 2, rank %d tensor received"
+                       % len(x.get_shape()))
+    return reduce_sum(array_ops.diag_part(x), name=name)
```
Comment (martinwicke) [inline tensorflow/python/ops/math_ops.py]: Sorry, one more, can you add another blank line after this function? We always have two blank lines between top-level functions.

File: tensorflow/python/ops/math_ops.py [hunk 5]
```diff
@@ -920,6 +921,39 @@ def reduce_any(input_tensor, reduction_indices=None, keep_dims=False,
                                                         reduction_indices),
                            keep_dims, name=name)
```
Comment (martinwicke) [inline tensorflow/python/ops/math_ops.py]: Another blank line here too.

File: tensorflow/python/kernel_tests/trace_op_test.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+  def testRankThreeFloatTensor(self):
+    x = numpy.random.rand(2, 2, 2)
+    self.assertRaises(ValueError, self.traceOp, x, numpy.float32, 0)
+
+  def testRankFourFloatTensor(self):
+    x = numpy.random.rand(2, 2, 2, 2)
+    self.assertRaises(ValueError, self.traceOp, x, numpy.float32, 0)
+
+if __name__ == "__main__":
```
Comment (martinwicke) [inline tensorflow/python/kernel_tests/trace_op_test.py]: One more blank line before if

**Comment by cameronphchen** [comment]
@martinwicke Thanks again for the review! Let me know if I need to squash these commits.

**Comment by martinwicke** [comment]
One more run of tests, then we're good.

Jenkins, test this please.

**Comment by martinwicke** [comment]
Merged.

**Comment by cameronphchen** [comment]
Thank you!

**Pull request closed without merging**